File: src/main/java/leetcode/topinterview/LeetCode131.java
package leetcode.topinterview;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/24 19:27
 */
public class LeetCode131 {

    //study
    //分两步
    boolean[][] dp;
    int n;
    List<List<String>> ans = new ArrayList<>();
    List<String> list = new ArrayList<>();

    public List<List<String>> partition(String s) {
        if (s == null || s.length() == 0) {
            return new ArrayList<>();
        }

        n = s.length();
        dp = new boolean[s.length()][s.length()];
        for (boolean[] d : dp) {
            Arrays.fill(d, true);
        }

        for (int i = n - 1; i >= 0; i--) {
            for (int j = i + 1; j < n; j++) {
                dp[i][j] = dp[i + 1][j - 1] && s.charAt(i) == s.charAt(j);
            }
        }
        dfs(s, 0);
        return ans;
    }

    private void dfs(String s, int i) {

        if (i == n) {
            ans.add(new ArrayList<>(list));
            return;
        }

        for (int j = i; j < n; j++) {
            if (dp[i][j]) {
                list.add(s.substring(i, j + 1));
                dfs(s, j + 1);
                list.remove(list.size() - 1);
            }
        }
    }
}

File: go/src/leetcode/topinterview/LeetCode131.go
package topinterview

func partition(s string) [][]string {
	if len(s) == 0 {
		return [][]string{}
	}
	dp := make([][]bool, len(s))
	for i := 0; i < len(s); i++ {
		dp[i] = make([]bool, len(s))
	}
	for i := 0; i < len(dp); i++ {
		for j := 0; j < len(dp[0]); j++ {
			dp[i][j] = true
		}
	}
	for i := len(s) - 1; i >= 0; i-- {
		for j := i + 1; j < len(s); j++ {
			dp[i][j] = dp[i+1][j-1] && s[i] == s[j]
		}
	}
	res := make([][]string, 0)
	var dfs func(s string, index int, list []string)
	dfs = func(s string, i int, list []string) {
		if i == len(s) {
			res = append(res, append([]string(nil), list...))
			return
		}
		for j := i; j < len(s); j++ {
			if dp[i][j] {
				list = append(list, s[i:j+1])
				dfs(s, j+1, list)
				list = list[:len(list)-1]
			}
		}
	}
	dfs(s, 0, []string{})
	return res
}
